File: tools/_registry.py
from dataclasses import dataclass
from typing import Callable, Generic, TypeVar
from pydantic import BaseModel
from tools._models import Example

In = TypeVar("In", bound=BaseModel)
Out = TypeVar("Out", bound=BaseModel)


@dataclass
class Tool(Generic[In, Out]):
    id: str
    name: str
    description: str
    category: str
    tags: list[str]
    methods: list[str]
    input_model: type[In]
    output_model: type[Out]
    compute: Callable[[In], Out]
    examples: list[Example]


REGISTRY: dict[str, Tool] = {}


def register(tool: Tool) -> Tool:
    if tool.id in REGISTRY:
        raise ValueError(f"duplicate tool id: {tool.id}")
    REGISTRY[tool.id] = tool
    return tool

File: packages/core/src/registry.ts
import type { z } from "zod";

export interface Tool<I = any, O = any> {
  id: string;
  name: string;
  description: string;
  category: string;
  tags: string[];
  methods: string[];
  // The third type param (the schema's INPUT) is left as `unknown` so schemas
  // that use .default()/nullable-with-default — whose parse-input type differs
  // from their z.output type — are still assignable. `I`/`O` track z.output.
  input: z.ZodType<I, z.ZodTypeDef, unknown>;
  output: z.ZodType<O, z.ZodTypeDef, unknown>;
  compute: (input: I) => O;
  examples: { input: unknown; output: unknown }[];
}

export const REGISTRY = new Map<string, Tool>();

export function register<I, O>(tool: Tool<I, O>): Tool<I, O> {
  if (REGISTRY.has(tool.id)) throw new Error(`duplicate tool id: ${tool.id}`);
  REGISTRY.set(tool.id, tool);
  return tool;
}
